modules/mapbox-layers/src/deck-utils.js

```diff
@@ -0,0 +1,113 @@
+import {Deck} from '@deck.gl/core';
+/* global window */
+
+export function getDeckInstance({map, gl}) {
+  // Only create one deck instance per context
+  if (map.__deck) {
+    return map.__deck;
+  }
+
+  const deck = new Deck({
+    // TODO - this should not be needed
+    canvas: 'deck-canvas',
+    width: '100%',
+    height: '100%',
+    controller: false,
+    useDevicePixels: true,
+    // layerFilter needs to be changed inside a rendering/picking cycle
+    // But calling setProps({layerFilter}) will trigger another rerender which sets off an infinite loop
+    // Instead, we use a constant callback here and access the dynamic filter in userData
+    layerFilter: ({layer}) => filterLayer(deck, layer),
+    _customRender: true,
+    userData: {
+      mapboxLayers: new Set()
+    }
+  });
+  deck._setGLContext(gl);
+  map.__deck = deck;
+
+  initAnimationLoop(map, deck);
+  initEvents(map, deck);
+
+  return deck;
+}
+
+export function addLayer(deck, layer) {
+  deck.props.userData.mapboxLayers.add(layer);
+  updateLayers(deck);
+}
+
+export function removeLayer(deck, layer) {
+  deck.props.userData.mapboxLayers.delete(layer);
+  updateLayers(deck);
+}
+
+export function updateLayer(deck, layer) {
+  updateLayers(deck);
+}
+
+export function drawLayer(deck, layer) {
+  deck.props.userData.layerFilter = layer.id;
+  deck._drawLayers('mapbox-repaint');
+  deck.needsRedraw({clearRedrawFlags: true});
+}
+
+function filterLayer(deck, layer) {
+  const {layerFilter} = deck.props.userData;
+
+  if (typeof layerFilter === 'string') {
+    return layer.id === layerFilter;
+  }
+  return layerFilter;
+}
+
+function updateLayers(deck) {
+  const layers = [];
+  deck.props.userData.mapboxLayers.forEach(deckLayer => {
+    const LayerType = deckLayer.props.type;
+    const layer = new LayerType(deckLayer.props);
+    layers.push(layer);
+  });
+  deck.setProps({layers});
+}
+
+// Create animation loop
+function initAnimationLoop(map, deck) {
+  let animationFrame = null;
+
+  function animate() {
+    deck.props.userData.layerFilter = false;
+    // Nothing's actually drawn yet because of the layer filter
+    // this just runs all the animation/layer updates then check if redraw flag is set
+    // TODO - pass luma-like animationProps
```

Comment:
I wonder if we should/could use the luma.gl `AnimationLoop` after all, some of the initial adaptations to deck.gl were based on the assumption that we wouldn't need deck.gl's animation loop. 

Did you see anythings that would need to be fixed to make that happen?

Having to emulate `animationProps` etc here seems fragile.

Reply:
That would be ideal. I will check the AnimationLoop class.